x/staking/keeper/msg_server.go

```diff
@@ -381,3 +384,112 @@ func (k msgServer) Undelegate(goCtx context.Context, msg *types.MsgUndelegate) (
+// and delegate back to validator
+func (k msgServer) CancelUnbondingDelegation(goCtx context.Context, msg *types.MsgCancelUnbondingDelegation) (*types.MsgCancelUnbondingDelegationResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	valAddr, err := sdk.ValAddressFromBech32(msg.ValidatorAddress)
+	if err != nil {
+		return nil, err
+	}
+	delegatorAddress, err := sdk.AccAddressFromBech32(msg.DelegatorAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	bondDenom := k.BondDenom(ctx)
+	if msg.Amount.Denom != bondDenom {
+		return nil, sdkerrors.Wrapf(
+			sdkerrors.ErrInvalidRequest, "invalid coin denomination: got %s, expected %s", msg.Amount.Denom, bondDenom,
+		)
+	}
+
+	// get validator
+	validator, found := k.GetValidator(ctx, valAddr)
+	if !found {
+		return nil, types.ErrNoValidatorFound
+	}
+
+	// In some situations, the exchange rate becomes invalid, e.g. if
+	// Validator loses all tokens due to slashing. In this case,
+	// make all future delegations invalid.
+	if validator.InvalidExRate() {
+		return nil, types.ErrDelegatorShareExRateInvalid
+	}
+
+	if validator.IsJailed() {
+		return nil, types.ErrValidatorJailed
+	}
+
+	ubd, found := k.GetUnbondingDelegation(ctx, delegatorAddress, valAddr)
+	if !found {
+		return nil, status.Errorf(
+			codes.NotFound,
+			"unbonding delegation with delegator %s not found for validator %s",
+			msg.DelegatorAddress, msg.ValidatorAddress)
+	}
+
+	var (
+		unbondEntry      types.UnbondingDelegationEntry
+		unbondEntryIndex int64 = -1
+	)
+
+	for i, entry := range ubd.Entries {
+		if entry.CreationHeight == int64(msg.CreationHeight) {
+			unbondEntry = entry
+			unbondEntryIndex = int64(i)
+			break
+		}
+	}
+	if unbondEntryIndex == -1 {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "unbonding delegation entry is not found at block height %d", msg.CreationHeight)
+	}
+
+	if !msg.Amount.IsPositive() {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "invalid amount")
+	}
+
+	if unbondEntry.Balance.LT(msg.Amount.Amount) {
+		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "amount is greater than the unbonding delegation entry balance")
+	}
+
+	if ubd.Entries[unbondEntryIndex].CompletionTime.Before(ctx.BlockTime()) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "unbonding delegation is already processed.")
+	}
+
+	// delegate the unbonding delegation amount to validator back
+	_, err = k.Keeper.Delegate(ctx, delegatorAddress, msg.Amount.Amount, types.Unbonding, validator, false)
+	if err != nil {
+		return nil, err
+	}
+
+	amount := unbondEntry.Balance.Sub(msg.Amount.Amount)
+	if amount.IsZero() {
+		// remove from ubd
+		ubd.RemoveEntry(unbondEntryIndex)
+	} else {
+		// update the unbondingDelegationEntryBalance and InitialBalance for ubd entry
+		ubd.Entries[unbondEntryIndex].Balance = amount
+		ubd.Entries[unbondEntryIndex].InitialBalance = ubd.Entries[unbondEntryIndex].InitialBalance.Sub(msg.Amount.Amount)
```

Comment:
Is it possible for `Balance` to be non-zero but `InitialBalance` be zero?

Reply:
No `InitialBalance`  never be zero 
`Balance` will be updated based on slashing factor of `InitialBalance` due to validator slashing

Reply:
If `Balance` is equal to Zero means delegator completely redelegate canceling  unbonding  amount back to his previous validator   so we will remove the `unbonding delegation entry` from `ubd Queue `

https://github.com/cosmos/cosmos-sdk/pull/10885/files/be0b721bc153a2504c3c8aea00eef541645e3e74#diff-5ea92aa6ab2a6d865b7641169f4a530e3ec4121e427e4bc292666fec403244daR468-R470